lib/rspec/core/configuration.rb

```diff
@@ -1597,6 +1597,39 @@ def define_derived_metadata(*filters, &block)
         @derived_metadata_blocks.append(block, meta)
       end
 
+      # Defines a callback that runs after the first example with matching
+      # metadata is defined. If no examples are defined with matching metadata,
+      # it will not get called at all.
+      #
+      # This can be used as an alternative to `before(:suite)` to define some
```

Comment:
> Did you consider options like `before_matching(:db, :suite)` etc?

I did not.  Actually, for `before(:suite)`, extra args are ignored (although we may print a warning -- I'd have to check), so we could potentially use the extra args here for this purpose.

I'm leaning against it, though.  Consider the use case I explained above -- where you want to conditionally `require 'support/db'` only if there are loaded examples matching `:db`.  If `support/db` also defines a `before(:suite)` hook, and the `require 'support/db'` line is executed by a `before(:suite)` hook -- then I'm not sure it'll work as desired.  We don't generally support you defining additional hooks while running said hooks.  I think we just get the list of hooks up front, and run them, without re-querying the hook state to see if there are new ones.

Or consider the main use case that drove this PR: the new `filter_run_when_matching` API we want to add.  Currently `before(:suite)` hooks are executed and example filtering is applied between all spec files being loaded and running the first spec, but I'm not sure which order they happen in.  Suite hooks might execute first, then filtering could be applied, or vice versa.  It hasn't been important before now and I don't think I want to make it important as part of this feature.

So I think I'd like to stick with `when_first_matching_example_defined` but I can rewrite the docs here to no longer compare it directly to `before(:suite)`.  I may re-write the example to show it being used to require a support file as well.

Would that be sufficient, @xaviershay?